### PR TITLE
Disable dynamic function dispatching for native or arch-specific builds

### PR DESCRIPTION
**Pull request by phprus**
Another implementation of PR #1645.

Main differences:
- Separate CPU-specific function prototypes and CPU-arch features checker.
- Reducing dependencies between header files.

### CODE REVIEW

File: arch/power/power_features.h [hunk 1]
```diff
@@ -1,11 +1,11 @@
-/* power_features.h -- check for POWER CPU features
+/* power_functions.h -- POWER implementations for arch-specific functions.
```
Comment (mtl1979) [inline arch/power/power_features.h]: Wrong filename

File: arch/power/power_functions.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/* power_features.h -- check for POWER CPU features
```
Comment (mtl1979) [inline arch/power/power_functions.h]: Wrong filename

File: arch/riscv/riscv_features.h [hunk 3]
```diff
@@ -1,39 +1,18 @@
-/* riscv_features.h -- check for riscv features.
+/* riscv_functions.h -- RISCV implementations for arch-specific functions.
```
Comment (mtl1979) [inline arch/riscv/riscv_features.h]: Wrong filename

File: arch/riscv/riscv_functions.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/* riscv_features.h -- check for riscv features.
```
Comment (mtl1979) [inline arch/riscv/riscv_functions.h]: Wrong filename

File: arch/arm/arm_functions.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#endif
+
+#ifdef ARM_SIMD
+void slide_hash_armv6(deflate_state *s);
+#endif
+
+
+#ifdef DISABLE_RUNTIME_CPU_DETECTION
+//// ARM - SIMD
+#  if (defined(ARM_SIMD) && 0 /* unknown predefined macro */ ) || defined(ARM_NOCHECK_SIMD)
```
Comment (mtl1979) [inline arch/arm/arm_functions.h]: ```__ARM_FEATURE_SIMD32```

Comment (phprus) [reply arch/arm/arm_functions.h]: Thank you!

**Comment by phprus** [comment]
All compile time CPU features checks is implemented.

**Comment by phprus** [comment]
Rebased.

**Comment by nmoinvaz** [comment]
I like how it is split into `cpu_features` and `cpu_functions`.

File: arch/arm/arm_functions.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+uint32_t update_hash_acle(deflate_state *const s, uint32_t h, uint32_t val);
+#endif
+
+#ifdef ARM_SIMD
+void slide_hash_armv6(deflate_state *s);
+#endif
+
+
+#ifdef DISABLE_RUNTIME_CPU_DETECTION
+//// ARM - SIMD
```
Comment (nmoinvaz) [inline arch/arm/arm_functions.h]: Noticed double `//`.

File: arch/arm/arm_functions.h [hunk 7]
```diff
@@ -0,0 +1,76 @@
+/* arm_functions.h -- ARM implementations for arch-specific functions.
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#ifndef ARM_FUNCTIONS_H_
+#define ARM_FUNCTIONS_H_
+
+
```
Comment (nmoinvaz) [inline arch/arm/arm_functions.h]: No need for extra line?

File: arch/generic/generic_functions.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+uint32_t compare256_unaligned_16(const uint8_t *src0, const uint8_t *src1);
+#  ifdef HAVE_BUILTIN_CTZ
+    uint32_t compare256_unaligned_32(const uint8_t *src0, const uint8_t *src1);
+#  endif
+#  if defined(UNALIGNED64_OK) && defined(HAVE_BUILTIN_CTZLL)
+    uint32_t compare256_unaligned_64(const uint8_t *src0, const uint8_t *src1);
+#  endif
+#endif
+
+typedef void    (*slide_hash_func)(deflate_state *s);
```
Comment (nmoinvaz) [inline arch/generic/generic_functions.h]: No extra space after `void`

File: arch/generic/generic_functions.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+#  define native_inflate_fast inflate_fast_c
+#  define native_insert_string insert_string_c
+#  define native_quick_insert_string quick_insert_string_c
+#  define native_slide_hash slide_hash_c
+#  define native_update_hash update_hash_c
+#  define native_longest_match longest_match_generic
+#  define native_longest_match_slow longest_match_slow_generic
+#  define native_compare256 compare256_generic
+#endif
+
```
Comment (nmoinvaz) [inline arch/generic/generic_functions.h:105]: Extra line?

File: functable.c [hunk 10]
```diff
@@ -2,11 +2,12 @@
  * Copyright (C) 2017 Hans Kristian Rosbach
  * For conditions of distribution and use, see copyright notice in zlib.h
  */
+#ifndef DISABLE_RUNTIME_CPU_DETECTION
 
 #include "zbuild.h"
-#include "zendian.h"
```
Comment (Dead2) [inline functable.c]: This dependency can be removed from win32/ makefiles. And the one below should be added.

**Comment by Dead2** [comment]
I don't see any of the new header files added to the win32/ makefiles, those should probably go in so the dependency handling works properly.
Actually we should probably make a script that generates those deps, and have a CI job that fails when there is a diff. Those are kind of a mess already, and steadily becoming worse. @mtl1979 what do you think, is it worth it or not?

I think perhaps the `cpu_functions.h` file should be named `arch_functions.h`, since it is not so much related to the running cpu (like `cpu_features.h` does) but rather to the arch-specific functions located in the `arch/*` directories.


I have only done a quick review so far, but what I have looked at so far looks good. I agree that the separation looks good.

**Comment by mtl1979** [comment]
> I don't see any of the new header files added to the win32/ makefiles, those should probably go in so the dependency handling works properly. Actually we should probably make a script that generates those deps, and have a CI job that fails when there is a diff. Those are kind of a mess already, and steadily becoming worse. @mtl1979 what do you think, is it worth it or not?
> 
> I think perhaps the `cpu_functions.h` file should be named `arch_functions.h`, since it is not so much related to the running cpu (like `cpu_features.h` does) but rather to the arch-specific functions located in the `arch/*` directories.
> 
> I have only done a quick review so far, but what I have looked at so far looks good. I agree that the separation looks good.

I have never looked up if cl.exe can generate dependency information for Windows. Even if it can, there's always issue that there is no sed-like tool to reformat the output so it matches the format nmake expects.

**Comment by nmoinvaz** [comment]
Time to ditch nmake like I've been saying for years. :)

**Comment by mtl1979** [comment]
> Time to ditch nmake like I've been saying for years. :)

nmake is still getting new features... It's just that some people use older versions, for example version that is bundled with Visual C++ 2017.9 ...

**Comment by nmoinvaz** [comment]
> nmake is still getting new features...

I can't locate a nmake change log anywhere..

File: functable.c (hunk 10 above)
Comment (phprus) [reply functable.c]: This dependency is not direct now, but indirect, through ``cpu_functions.h``.

As I see the dependencies in ``win32/Makefile.*`` were broken before this PR.
For example, ``win32/Makefile.arm`` has a dependency on x86 file ``$(SRCDIR)/arch/x86/x86_features.h``.

If we need actual nmake support (without cmake), I suggest fix the nmake-dependencies with separate PRs.

**Comment by phprus** [comment]
Rebased.

@Dead2, Renamed cpu_functions.h to arch_functions.h.

**Review by nmoinvaz — APPROVED** [review]

**Comment by Dead2** [comment]
@phprus I'd suggest splitting the last commit "Allow overwrite NATIVEFLAG value" into a separate PR. It increases the complexity of an already complex PR, and would benefit from getting reviewed and tested separately, and it'll avoid delaying this PR.

**Comment by phprus** [comment]
@Dead2 
I added commit "Allow overwrite NATIVEFLAG value" so that I could force setup the instruction set and check that the build was correct.
So in AVX build only SSE extensions should be used.

We can separate it in other PR and merge it before this PR. After merge I can rebase this over new commit.

**Comment by phprus** [comment]
PR #1662

**Comment by phprus** [comment]
Rebased and ready for review.

File: arch/power/power_features.h [hunk 11]
```diff
@@ -1,11 +1,11 @@
-/* power_features.h -- check for POWER CPU features
+/* power_features.h -- POWER implementations for arch-specific functions.
```
Comment (mtl1979) [inline arch/power/power_features.h]: Unnecessary change.

File: arch/power/power_functions.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
+/* power_functions.h -- check for POWER CPU features
```
Comment (mtl1979) [inline arch/power/power_functions.h]: Wrong description, see `power_features.h`

File: arch/riscv/riscv_features.h [hunk 13]
```diff
@@ -1,39 +1,18 @@
-/* riscv_features.h -- check for riscv features.
+/* riscv_features.h -- RISCV implementations for arch-specific functions.
```
Comment (mtl1979) [inline arch/riscv/riscv_features.h]: Unnecessary change.

File: arch/riscv/riscv_functions.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
+/* riscv_functions.h -- check for riscv features.
```
Comment (mtl1979) [inline arch/riscv/riscv_functions.h]: Wrong description, see `riscv_features.h`.

File: arch/s390/s390_functions.h [hunk 15]
```diff
@@ -0,0 +1,16 @@
+#ifndef S390_FUNCTIONS_H_
```
Comment (mtl1979) [inline arch/s390/s390_functions.h]: Missing the comment mentioning what the file does as seen on similar files. All new files should mention that the code is licensed under zlib license.

**Review by mtl1979 — CHANGES_REQUESTED** [review]
Needs cleanup, some unnecessary changes, missing reference to license in new files.

**Comment by mtl1979** [comment]
> > nmake is still getting new features...
> 
> I can't locate a nmake change log anywhere..

I just compare the documentation between different Visual Studio versions... As I have 2017.9, 2019 and 2022 installed on same PC, I can also just test if some "newer" feature isn't supported in earlier versions.

**Comment by phprus** [comment]
(Un)installing dependencies on Windows failed:

```
Chocolatey v2.2.2
Uninstalling the following packages:
strawberryperl

strawberryperl v5.32.1.1
 Running auto uninstaller...
 Auto uninstaller has successfully uninstalled strawberryperl or detected previous uninstall.
 strawberryperl has been successfully uninstalled.
Environment Vars (like PATH) have changed. Close/reopen your shell to
 see the changes (or in powershell/cmd.exe just type `refreshenv`).

Chocolatey uninstalled 1/1 packages. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
Chocolatey v2.2.2
Installing the following packages:
ninja
By installing, you accept licenses for the packages.
Invalid credentials specified.
Please provide credentials for: https://community.chocolatey.org/api/v2/
User name: Password: Chocolatey had an error on fv-az731-175 (with user runneradmin):
System.InvalidOperationException: Cannot read keys when either application does not have a console or when console input has been redirected from a file. Try Console.Read.
   at System.Console.ReadKey(Boolean intercept)
   at chocolatey.infrastructure.commandline.ReadKeyTimeout.ConsoleReadKey()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

Unhandled Exception: System.InvalidOperationException: Cannot read keys when either application does not have a console or when console input has been redirected from a file. Try Console.Read.
   at System.Console.ReadKey(Boolean intercept)
   at chocolatey.infrastructure.commandline.ReadKeyTimeout.ConsoleReadKey()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
Error: Process completed with exit code 1.

```

File: functable.c (hunk 10 above)
Comment (mtl1979) [reply functable.c]: I can try to fix the nmake dependencies when all the "more important" changes have been merged... It would help a lot if people would create issues when there is something wrong in the nmake Makefiles as I can't track all the changes if pull requests are merged in batches. I try to check most of the new pull requests at least once a week (unless I'm tagged in the pull request), but my time is limited due to other projects that are more important.

**Comment by mtl1979** [comment]
> Cannot read keys when either application does not have a console or when console input has been redirected from a file.

The issue is still unresolved after months, see https://github.com/chocolatey/choco/issues/2805

**Comment by phprus** [comment]
Ping?

**Comment by Dead2** [comment]
> Ping?

I am currently reviewing it, again. It is a big structural change, so it is a complicated PR to review.
In retrospect, it would have been easier if it was two PRs, one for reorganizing the code into the new file hierarchy but with no changes to output, and a separate one for the new feature disabling functable.

**Comment by Dead2** [comment]
# Develop ca0e4634e1914ecfe67e40261bbf54676f0d1c8f Feb 21
```
 Tool: minigzip   Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.150/0.154/0.156/0.001        0.034/0.036/0.038/0.001        8,523,732
 2     43.871%      0.244/0.249/0.250/0.001        0.032/0.036/0.038/0.002        6,903,609
 3     42.387%      0.295/0.297/0.299/0.001        0.031/0.034/0.036/0.001        6,670,099
 4     41.647%      0.317/0.324/0.326/0.002        0.030/0.034/0.035/0.001        6,553,723
 5     41.216%      0.344/0.352/0.354/0.002        0.032/0.034/0.035/0.001        6,485,938
 6     41.037%      0.390/0.397/0.400/0.003        0.030/0.033/0.035/0.001        6,457,776
 7     40.778%      0.493/0.498/0.503/0.003        0.030/0.034/0.035/0.001        6,416,919
 8     40.704%      0.605/0.613/0.618/0.004        0.029/0.033/0.034/0.001        6,405,244
 9     40.409%      0.912/0.919/0.925/0.004        0.029/0.033/0.035/0.002        6,358,951

 avg1  42.913%                        0.423                          0.034
 tot                                114.095                          9.195       60,775,991

   text    data     bss     dec     hex filename
 138190    1352       8  139550   2211e libz-ng.so.2
``` 

 ## Develop Native=on
 ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.147/0.151/0.153/0.002        0.028/0.035/0.037/0.002        8,523,732
 2     43.871%      0.239/0.242/0.244/0.001        0.031/0.035/0.037/0.002        6,903,609
 3     42.387%      0.285/0.289/0.291/0.002        0.031/0.034/0.035/0.001        6,670,099
 4     41.647%      0.316/0.318/0.320/0.001        0.030/0.033/0.035/0.001        6,553,723
 5     41.216%      0.345/0.351/0.354/0.002        0.030/0.033/0.034/0.001        6,485,938
 6     41.037%      0.389/0.397/0.401/0.003        0.029/0.033/0.034/0.002        6,457,776
 7     40.778%      0.485/0.495/0.498/0.003        0.030/0.033/0.035/0.001        6,416,919
 8     40.704%      0.600/0.607/0.611/0.003        0.027/0.032/0.034/0.002        6,405,244
 9     40.409%      0.903/0.909/0.912/0.003        0.028/0.033/0.034/0.002        6,358,951

 avg1  42.913%                        0.418                          0.033
 tot                                112.795                          9.000       60,775,991

   text    data     bss     dec     hex filename
 137222    1352       8  138582   21d56 libz-ng.so.2
``` 
 ## PR #1659 Native=off
 ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.151/0.154/0.155/0.001        0.031/0.035/0.037/0.001        8,523,732
 2     43.871%      0.247/0.249/0.250/0.001        0.032/0.036/0.038/0.002        6,903,609
 3     42.387%      0.291/0.295/0.297/0.002        0.032/0.035/0.036/0.001        6,670,099
 4     41.647%      0.319/0.323/0.326/0.002        0.032/0.034/0.035/0.001        6,553,723
 5     41.216%      0.345/0.351/0.354/0.002        0.029/0.034/0.035/0.001        6,485,938
 6     41.037%      0.392/0.397/0.400/0.002        0.031/0.033/0.035/0.001        6,457,776
 7     40.778%      0.491/0.499/0.503/0.003        0.029/0.033/0.035/0.001        6,416,919
 8     40.704%      0.606/0.611/0.616/0.003        0.031/0.033/0.035/0.001        6,405,244
 9     40.409%      0.911/0.918/0.924/0.004        0.031/0.033/0.035/0.001        6,358,951

 avg1  42.913%                        0.422                          0.034
 tot                                113.946                          9.209       60,775,991

   text    data     bss     dec     hex filename
 138190    1352       8  139550   2211e libz-ng.so.2
```
## PR #1659 Native=on
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.093/0.097/0.099/0.001        0.027/0.034/0.037/0.003        8,523,732
 2     43.871%      0.172/0.176/0.178/0.002        0.032/0.035/0.037/0.001        6,903,609
 3     42.387%      0.224/0.227/0.229/0.001        0.031/0.034/0.035/0.001        6,670,099
 4     41.647%      0.253/0.256/0.259/0.002        0.029/0.033/0.035/0.002        6,553,723
 5     41.216%      0.285/0.290/0.293/0.002        0.029/0.033/0.034/0.002        6,485,938
 6     41.037%      0.328/0.338/0.342/0.003        0.028/0.032/0.034/0.002        6,457,776
 7     40.778%      0.489/0.495/0.499/0.003        0.027/0.033/0.035/0.002        6,416,919
 8     40.704%      0.601/0.608/0.612/0.003        0.028/0.033/0.034/0.002        6,405,244
 9     40.409%      0.903/0.912/0.918/0.004        0.025/0.031/0.033/0.002        6,358,951

 avg1  42.913%                        0.378                          0.033
 tot                                102.026                          8.896       60,775,991

   text    data     bss     dec     hex filename
 133814    1208       8  135030   20f76 libz-ng.so.2
 ```
 
 Develop w/native=ON vs PR w/native=ON compression speedup:
 * Level 1: 35%
 * Level 3: 21%
 * Level 6: 15%
 * Level 9: 0.3%
 
 PR w/native=ON vs PR #1681 w/native=OFF:
 * Level 1: 11%
 * Level 3: 8%
 * Level 6: 5%
 * Level 9: 0.5% slower

Looks like a decent speedup to me.
Ideally this would have been based on top of #1681, as it removes a good portion of the complexity in this PR.

**Comment by phprus** [comment]
My suggestion:

1) Extract commits:

- Split CPU features checks and CPU-specific function prototypes and reduce include-dependencies.
- Move select for generic functions into generic_functions.h.
- Rename cpu_functions.h to arch_functions.h.

into separate PR.

2) Rebase #1681 onto separated files implementation.

3) Rebase this PR onto new develop branch after previous PRs is merged.


What do your think about this plan?

**Comment by Dead2** [comment]
> What do your think about this plan?

@phprus I think that is a good plan and would probably ensure we get it all merged the fastest. 👍

**Comment by phprus** [comment]
@Dead2 

1) PR #1684 - small fix for #1662.
2) PR #1685 - split headers.

**Comment by phprus** [comment]
~~Rebase required.~~
~~Waiting PR #1673 (NMake)~~
~~Waiting #1681, #1682 (deflate optimizations).~~

**Comment by phprus** [comment]
Ready for review.

File: .github/workflows/cmake.yml [hunk 16]
```diff
@@ -516,13 +521,23 @@ jobs:
...
+            cmake-args: -T ClangCl -A x64 -DWITH_NATIVE_INSTRUCTIONS=ON -DNATIVE_ARCH_OVERRIDE="-mavx -mpclmul"
+
           - name: Windows GCC
             os: windows-latest
             compiler: gcc
             cxx-compiler: g++
             cmake-args: -G Ninja
             codecov: win64_gcc
 
+          - name: Windows GCC (AVX)
```
Comment (Dead2) [inline .github/workflows/cmake.yml]: **nitpick:** 
This should also have `Native Instructions` in the name.

File: deflate.c [hunk 17]
```diff
@@ -190,8 +190,10 @@ int32_t ZNG_CONDEXPORT PREFIX(deflateInit2)(PREFIX3(stream) *strm, int32_t level
     deflate_state *s;
     int wrap = 1;
 
+#ifndef DISABLE_RUNTIME_CPU_DETECTION
     /* Force initialization functable, because deflate captures function pointers from functable. */
     functable.force_init();
+#endif
```
Comment (Dead2) [inline deflate.c]: **suggestion:** 
I am wondering whether it'd be cleaner and more in theme with the rest of it like `FUNCTABLE_CALL` etc if this instead was replaced by something like this:
```
    /* Initialize functable */
    FUNCTABLE_INIT;
```

The current comment is correct, but it is also a little misleading, since we don't only do it because deflate captures function pointers. It is a good idea to do the init work in the init functions, so that applications can hide that latency if they need to.

File: inflate.c [hunk 18]
```diff
@@ -139,8 +139,10 @@ int32_t ZNG_CONDEXPORT PREFIX(inflateInit2)(PREFIX3(stream) *strm, int32_t windo
     int32_t ret;
     struct inflate_state *state;
 
+#ifndef DISABLE_RUNTIME_CPU_DETECTION
     /* Initialize functable earlier. */
     functable.force_init();
+#endif
```
Comment (Dead2) [inline inflate.c]: **suggestion:** 
Same here.

File: .github/workflows/cmake.yml (hunk 16 above)
Comment (phprus) [reply .github/workflows/cmake.yml]: Fixed

File: deflate.c (hunk 17 above)
Comment (phprus) [reply deflate.c]: Done.

File: inflate.c (hunk 18 above)
Comment (phprus) [reply inflate.c]: Done.

**Review by Dead2 — APPROVED** [review]
LGTM

**Comment by phprus** [comment]
Rebased onto develop

File: .github/workflows/cmake.yml [hunk 19]
```diff
@@ -546,6 +561,12 @@ jobs:
             gcov-exec: gcov-10
             codecov: macos_gcc
 
+          - name: macOS Clang Native Instructions (ARM64)
+            os: macos-14
```
Comment (nmoinvaz) [inline .github/workflows/cmake.yml:565]: Can we use `macos-latest` here instead? If there is a reason, perhaps it should be commented?

**Review by nmoinvaz — APPROVED** [review]

File: .github/workflows/cmake.yml (hunk 19 above)
Comment (phprus) [reply .github/workflows/cmake.yml:565]: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners

> The macos-latest workflow label currently uses the macOS 12 runner image.

Therefore we need macos-14 (first ARM64 runner).

Comment (nmoinvaz) [reply .github/workflows/cmake.yml:565]: I see. Got it.
